KAFKA-15713: KRaft support in AclCommandTest

pasharik opened the pull request:
- Updating `AclCommandTest` to support KRaft
- Tests which are using `AclAuthoriser` are not updated, as they are expected to be removed after full migration to KRaft
- Changed `AclCommand` console output for adding and removing ACLs.
In the original implementation, `listAcls()` method was called directly from `addAcls()` and `removeAcls()` methods, which caused a race condition in KRaft mode, so the test become flaky
  - This is quite a big change for our UI, as I understand. Should it go throuhg KIP process?

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

chia7712 commented on core/src/main/scala/kafka/admin/AclCommand.scala:
```diff
@@ -115,8 +115,6 @@ object AclCommand extends Logging {
           adminClient.createAcls(aclBindings).all().get()
         }
-
-        listAcls(adminClient)
```
why removing this print?

pasharik replied:
Restored this print for now.

For me, this print together with checking for the output of this print inside the test, causes kraft tests to be flaky. E.g.
```
./gradlew core:test --tests "kafka.admin.AclCommandTest.testAclCliWithAdminAPI" --rerun
```

it always passes for `zk`, but quite often fails for `kraft`.

As I understand, `listAcls()` can be served by any broker in kraft mode, not just by controller. So potentially some brokers may not have most up-to-date copy of metadata right after `createAcls()` call. Here there is no interval between `createAcls()` and `listAcls()`, so probability of such race condition is higher.
From [KIP-500](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500) and the [diagram](https://cwiki.apache.org/confluence/download/attachments/123898922/b.png) it seems what brokers have to periodically pull new metadata from the kraft controller, so metadata is not available on all brokers immediately.
Please correct me if I'm wrong here.

I've looked at `TopicCommand`, as another example of the class which manages metadata. It seems what it doesn't invoke `listTopics()` after creating or deleting topics. So I thought we can use similar approach here.

I tried to reproduce this flaky behavior with running Kafka controller and broker locally with `bin/kafka-server-start.sh`, and invoking `bin/kafka-acls.sh` manually, but wasn't able to reproduce same issue.

Probably, it can be related to the test infrastructure setup. I'll try to re-write the test to java and use new test infrastructure, let's see it solves the issue

pasharik replied:
Restored this print for now.

For me, this print together with checking for the output of this print inside the test, causes kraft tests to be flaky. E.g.
```
./gradlew core:test --tests "kafka.admin.AclCommandTest.testAclCliWithAdminAPI" --rerun
```

it always passes for `zk`, but quite often fails for `kraft`.

As I understand, `listAcls()` can be served by any broker in kraft mode, not just by controller. So potentially some brokers may not have most up-to-date copy of metadata right after `createAcls()` call. Here there is no interval between `createAcls()` and `listAcls()`, so probability of such race condition is higher.
From [KIP-500](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500) and the [diagram](https://cwiki.apache.org/confluence/download/attachments/123898922/b.png) it seems what brokers have to periodically pull new metadata from the kraft controller, so metadata is not available on all brokers immediately.
Please correct me if I'm wrong here.

I've looked at `TopicCommand`, as another example of the class which manages metadata. It seems what it doesn't invoke `listTopics()` after creating or deleting topics. So I thought we can use similar approach here.

I tried to reproduce this flaky behavior with running Kafka controller and broker locally with `bin/kafka-server-start.sh`, and invoking `bin/kafka-acls.sh` manually, but wasn't able to reproduce same issue.

Probably, it can be related to the test infrastructure setup. I'll try to re-write the test to java and use new test infrastructure, let's see it solves the issue

chia7712 replied:
> As I understand, listAcls() can be served by any broker in kraft mode, not just by controller. So potentially some brokers may not have most up-to-date copy of metadata right after createAcls() call. Here there is no interval between createAcls() and listAcls(), so probability of such race condition is higher.
From [KIP-500](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500) and the [diagram](https://cwiki.apache.org/confluence/download/attachments/123898922/b.png) it seems what brokers have to periodically pull new metadata from the kraft controller, so metadata is not available on all brokers immediately.
Please correct me if I'm wrong here.

agree to race condition. 

> Probably, it can be related to the test infrastructure setup. I'll try to re-write the test to java and use new test infrastructure, let's see it solves the issue

agree that we should fix the test infra instead of changing the output of command tool. I worried about that some users may rely on the output of `listAcls`, so we should not change the output arbitrarily.

pasharik replied:
- I've moved KRaft tests into a new `AclCommandIntegrationTest.java`
- Left old Zookeeper tests in `AclCommandTest.scala`. As I understand, we are going to completely delete this test file, once fully moved to KRaft, am I right? Do you think it's worth migrating those tests to java at this stage, if they are going to be deleted anyway?
- Race condition described above, is still reproduced on a new infrastructure :cry: So if there are no objections, we can probably remove console output `Current ACLs`

chia7712 replied:
> Do you think it's worth migrating those tests to java at this stage, if they are going to be deleted anyway?

we can keep origin test in this PR. the new test file you added is good to me

> we can probably remove console output Current ACLs

agree

chia7712 reviewed (commented):
@pasharik thanks for this contribution. We are migrating the command tools from scala to java, and hence could you rewrite it by java? #15645 is a good example. Also, using the new test infra will be great :)

soarez reviewed (commented):
Hi @pasharik. Thanks for the change.


> In the original implementation, listAcls() method was called directly from addAcls() and removeAcls() methods, which caused a race condition in KRaft mode, so the test become flaky

Can you share a bit more detail on this? How specifically does `listAcls()` contribute to flakiness? Thanks

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest - restoring listAcls() in AclCommand

pasharik commented:
> Hi @pasharik. Thanks for the change.
> 
> > In the original implementation, listAcls() method was called directly from addAcls() and removeAcls() methods, which caused a race condition in KRaft mode, so the test become flaky
> 
> Can you share a bit more detail on this? How specifically does `listAcls()` contribute to flakiness? Thanks

Replied in the [above thread](https://github.com/apache/kafka/pull/15830#discussion_r1584190323)

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

chia7712 commented:
@pasharik Could you please check  the failed test?

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

Removing kraft support from the following tests:
* testAclCliWithClientId
* testAclCliWithAdminAPI
* testProducerConsumerCliWithAdminAPI

chia7712 commented on core/src/test/scala/unit/kafka/admin/AclCommandTest.scala:
```diff
@@ -122,19 +128,27 @@ class AclCommandTest extends QuorumTestHarness with Logging {
-  @Test
-  def testAclCliWithAdminAPI(): Unit = {
+  @ParameterizedTest
+  @ValueSource(strings = Array("zk"))
```
Why this test case can't run with kraft?

pasharik replied:
Some tests are failing in Kraft mode because of a race condition, which we [discussed above](https://github.com/apache/kafka/pull/15830#discussion_r1585713585)

Currently I'm trying to rewrite the whole test in java and migrate to a new test infrastructure. Hopefully, it will solve the race condition issue. If it doesn't help, then probably we'll have to remove some failing checks from the test or alternatively change the output of the AclCommand tool itself.

I've removed kraft support from those test for now, just to make a jenkins build pass. Planning to remove this scala test class complately, and replace if with a new java test class

chia7712 replied:
Maybe we can create `AclCommand.AdminClientService` to test its method `listAcls` with retry instead of grabbing the output from `main` ?

pasharik replied:
Commited [alternative fix](https://github.com/apache/kafka/pull/15830/commits/d39be8a7c349e868b7c5066e6be89f14a9e6adf3) for failing tests

For now, I've added a condition, so we don't verify the output from the `main` for the Kraft tests. For adding ACLs:

```
if (!isKRaftTest()) {
  assertOutputContains(...)
```

and for deleting ACLs:
```
if (!isKRaftTest()) {
  assertEquals("", out)
```

Just wondering, do we need to check this `main` output at all?

> Maybe we can create `AclCommand.AdminClientService` to test its method `listAcls` with retry instead of grabbing the output from `main` ?

Yeah, there is already similar logic in the test:

```
for (resource <- resources) {
  withAuthorizer() { authorizer =>
    TestUtils.waitAndVerifyAcls(...)
```

so eventualy ACLs are propageted in Kraft mode, and tests pass.

The only problem is with capturing the immediate output from the `main`, which may not be correct sometimes, because of the race condigion

chia7712 replied:
> Just wondering, do we need to check this main output at all?

from the testing we can know that output "Current ACLs" could be out-of-date information. +1 to your previous changes that removing `listAcls` from `removeAcls` and `addAcls` in order to avoid confusing users.

chia7712 reviewed (commented):
@pasharik thanks for updated PR. I notice that some test cases don't run with kraft. Could you please add comments to say why this test case can't run with kraft?

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

Restoring kraft support from the following tests:
* testAclCliWithClientId
* testAclCliWithAdminAPI
* testProducerConsumerCliWithAdminAPI

Removing checking the output from main for the Kraft tests

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

chia7712 commented on core/src/test/java/kafka/admin/AclCommandIntegrationTest.java:
```diff
@@ -0,0 +1,453 @@
+    private static final String SUPER_USERS = "super.users";
+    private static final String USER_ANONYMOUS = "User:ANONYMOUS";
+
+    private static final KafkaPrincipal PRINCIPAL = SecurityUtils.parseKafkaPrincipal("User:test2");
```
ok, that will have many duplicate code ...

Sorry that could we merge this test into `AclCommandTest`? We don't need to use new test infra in this PR, and we can do a follow-up for that. Also, using scala is ok to me for now

pasharik replied:
Okay, restored it back:
- Moved KRaft tests back into `AclCommandTest.scala`
- Deleted `AclCommandIntegrationTest.java`
- Removed `"Current ACLs"` print output from the AclCommand itself

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

chia7712 reviewed (approved):
LGTM

chia7712 commented:
@pasharik Could you please take a look at failed tests?

pasharik committed:
Trigger build

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

pasharik commented:
> Could you please take a look at failed tests?

`AclCommandTest` seems to be fixed, but there are some other failing tests for jdk8 and jdk21

chia7712 commented on core/src/test/scala/unit/kafka/admin/AclCommandTest.scala:
```diff
@@ -325,11 +349,15 @@ class AclCommandTest extends QuorumTestHarness with Logging {
+    val auth = if (isKRaftTest()) {
+      servers.last.authorizer.get
+    } else {
+      new AclAuthorizer
```
Maybe we can use the authorizer of broker instead of creating new one?

pasharik replied:
As I understand, this AclAuthorizer is used for some tests like `testAclCliWithAuthorizer`, `testProducerConsumerCliWithAuthorizer`, `testAclsOnPrefixedResourcesWithAuthorizer`, `testPatternTypesWithAuthorizer`. Those tests don't use AdminClient, so for them there seems to be no broker created at all, only Zookeeper instance. So there is no broker/controller from which we can obtain the AclAutorizer, hence creating a new instance of the autorizer here

chia7712 reviewed (commented):
@pasharik Thanks for updated PR. I loop `testAclCliWithClientId` and it is flaky, and so could you please take a look?

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

pasharik commented:
Test seems to be more stable now after recent update. I'm also running it in a loop in IntelliJ, and so far can't reproduce flakyness :crossed_fingers:

chia7712 commented on core/src/test/scala/unit/kafka/admin/AclCommandTest.scala:
```diff
@@ -324,12 +348,18 @@ class AclCommandTest extends QuorumTestHarness with Logging {
-      f(authZ)
-    } finally authZ.close()
+    if (isKRaftTest()) {
+      (servers.map(_.authorizer.get) ++ controllerServers.map(_.authorizer.get)).foreach { auth =>
```
It seems to me those code works well with zk, so we can simplify `withAuthorizer`.
```scala
  private def withAuthorizer()(f: Authorizer => Unit): Unit = {
    (servers.map(_.authorizer.get) ++ controllerServers.map(_.authorizer.get)).foreach { auth =>
      f(auth)
    }
  }
```

pasharik replied:
I've simplified the `withAuthorizer()` method with the implementation above. But there seems to be an issue with test methods like `testAclCliWithAuthorizer()`, `testProducerConsumerCliWithAuthorizer()`, `testAclsOnPrefixedResourcesWithAuthorizer()`, `testPatternTypesWithAuthorizer()`. In those methods, there are no running server or controller, so the above code in `withAuthorizer()` just does not execute anything at all, because the collection `(servers.map(_.authorizer.get) ++ controllerServers.map(_.authorizer.get))` is empty. 

Those tests pass, but `withAuthorizer()` doesn't check anything for them. So probably we are getting a bit less code coverege in this case.

Not sure if it's a big issue, as we are going to delete those test methods after full migration to KRaft anyway?

chia7712 replied:
> Those tests pass, but withAuthorizer() doesn't check anything for them. So probably we are getting a bit less code coverege in this case.

that is a good point. in zk mode we can modify the data through zk client, and hence it can work even though there is no broker. In order to test authorizer by interface (`KafkaBroker`), maybe we can create a broker for those tests after changing the data of zk. WDYT?

pasharik replied:
Would it be a bit confusing for someone who looks at `testAclCliWithAuthorizer` and similar tests, if we start broker just for verifying ACLs in Zk?
For me, the original implementation of `withAuthorizer()` with extra check `if (isKRaftTest())` looks fine... at least it would be easy to remove the `else` branch, once we get rid of Zk - we'll just look for all places where `isKRaftTest()` is used in this test class

To optimise it a bit, maybe we can create and configure this `new AclAuthorizer` when test starts, e.g. in `setUp()` method. Something like this:

```
setUp() { ...
  aclAutorizerForZk = new AclAuthorizer
  aclAutorizerForZk.configure(..)
}

private def withAuthorizer()(f: Authorizer => Unit): Unit = {
    if (isKRaftTest()) {
      (servers.map(_.authorizer.get) ++ controllerServers.map(_.authorizer.get)).foreach { auth =>
        f(auth)
      }
    } else {
        f(aclAutorizerForZk)
    }
}
```

chia7712 replied:
My point was the test should honor `KafkaBroker` interface instead of creating/configuring the authorizer manually. However, your comment makes sense to me that zk is going to be dropped. My origin idea is over-engineering.

> To optimise it a bit, maybe we can create and configure this new AclAuthorizer when test starts, e.g. in setUp() method. Something like this:

That will create zk authorizer even though the tests don't have zk, right? 

> the original implementation of withAuthorizer() with extra check if (isKRaftTest()) looks fine... at least it would be easy to remove the else branch, once we get rid of Zk - we'll just look for all places where isKRaftTest() is used in this test class

let's keep original implementation :)

pasharik replied:
> That will create zk authorizer even though the tests don't have zk, right?

Yeah, you are right...

Restored the original implementation :ok_hand:

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

pasharik committed:
KAFKA-15713: KRaft support in AclCommandTest

chia7712 commented:
@pasharik thanks for your contribution and patience :)
